### PR TITLE
Enumerable Methods

### PR DESCRIPTION
**Pull request by Div685**
This project is based on the Odin Project of Enumerable methods.
We created our own methods on to the existing Enumerable module without using original methods and with yield statement or #call method as below:

- #my_each
- #my_each_with_index
- #my_select
- #my_all?
- #my_any?
- #my_none?
- #my_count
- #my_map
- #my_inject
- #multiply_els

#multiply_els is created for testing #my_inject, which multiplies all the elements of the array together by using #my_inject.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,106 @@
-# enumerables
\ No newline at end of file
+# Enumerable Methods
+
+![](https://img.shields.io/badge/Microverse-blueviolet)
+
+## About the Project
+
+![Top Page Screenshot](./screenshot.jpg)
```
Comment (arslanbisharat) [inline README.md]: - [x] I cannot see any screenshot in your project. Please have another look

File: enumerables.rb [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    array.size.times { |index| yield(array[index]) }
+    self
+  end
+
+  def my_each_index
+    return enum_for(:my_each) unless block_given?
+
+    array = *self
+    array.size.times { |index| yield(index) }
+  end
```
Comment (arslanbisharat) [inline enumerables.rb]: - [x] It should have working `my_each_with_index` identical and returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

#### my_each_with_index

- [x] my_each_with_index returns an Enumerator if no block is given

- Yours is always expecting a block
image

- Real each_with_index
image

File: enumerables.rb [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+  def my_map
+    return enum_for(:my_each) unless block_given?
+
+    arr = *self
+    new_array = []
+    arr.my_each { |item| new_array << yield(item) }
+    new_array
+  end
```
Comment (arslanbisharat) [inline enumerables.rb]: * [x] should have working `my_map` identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

#### my_map 

- [x] returns an Enumerator if no block is given
```ruby
[2,4,7,11].my_map should be equal to [2,4,7,11].map #<Enumerator: [2, 4, 7, 11]:map
```
- Yours is always expecting for a block
image

- Real map
image

- [x] in the specifications of the project in the [Odin project](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks) # 12, you were asked to modify your my_map method to take either a proc or a block. And if both a proc and a block are given, only execute the proc. The real map doesn't have this feature, but you were asked to have it so I strongly recommend working on it

Given 
```ruby
my_proc = Proc.new {|num| num > 10 }
```
Result
```ruby
puts [18, 22, 5, 6] .my_map(my_proc) {|num| num < 10 }  => true true false false
```

**Review by arslanbisharat — CHANGES_REQUESTED** [review]
### Request changes 😳
You have done a great job so far. Your project is almost completed. But I found 8 errors(Tests are falling) in your project.
 I have given some suggestions below. Please work on them. 
_It's okay. We are here to learn. Don't get discouraged. Work even harder. You are almost there.😄_

**Please make the changes and request re-review for the given project**
Stay safe and give me a smile 😇

### Happy Coding
---------------------------------------------------------------------------------------------------------------------------------
**Regards,
Arslan Bisharat
slack: arslanbisharat**

File: enumerables.rb (hunk 3 above)
Comment (yokoworks) [reply enumerables.rb]: We tried to do as you mentioned above, but it doesn't work on our local PC.
Could you tell us specifically how we should run our new my_map method with proc?

Screen Shot 2020-11-03 at 23 51 06

Comment (arslanbisharat) [reply enumerables.rb]: I hope this [article ](https://docs.ruby-lang.org/en/2.6.0/Proc.html) might help you.

Comment (yokoworks) [reply enumerables.rb]: We just fixed it, and it should work now. 
Actually, on our local PC, it's working now.
Thanks a lot.

**Review by layzrd — CHANGES_REQUESTED** [review]
### Status: Require changes  ♲
Hi @Div685  and @yocosaka, 👏 You did very well on this project. ✨ Big-up for successfully implementing the enumerable methods. Most requirements have done 👌, Kindly fix the following issues.

[REQUIRED]

- [x] The my_map method should execute only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.

Once you fix the issues, kindly submit another code review. ♲

Happy coding 👨‍💻
[Abdulaziz Redi Ali](https://azizali.ml/)

**Review by mahmoud717 — APPROVED** [review]
### Project approved ✔️ ✔️ 
**Hi @yocosaka & @Div685**
great job implementing the required changes.
the project is approved, you can go ahead and merge this pull request.

**Pull request merged**